Separate game rule enabling and game rule starting

mirrorcult opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Needed for Ship v Ship gamemode, as that cares about map loading which before this PR happened way before gamerule enabling.

Now, game rules are enabled in PreRoundSetup, and Started (override functions called) when the round starts, to avoid conflicts. This way gamerules can care about map load for events but not actually do anything specific until the round starts.

ElectroJr reviewed (commented):
So maybe I'm getting somewhat confused, but as I understand:
- If a rule is enabled/disabled it gets added to `_gameRules`
- After being enabled/disabled, a rule can be started/stopped

The terminology is a bit weird, and I think something like add/remove+start/stop or add/remove+enable/disable make more sense than enable/disable+start/stop

I assume you might've been avoiding reusing add/remove to avoid confusion with the old add/remove functions, but it makes it even more confusing IMO.

ElectroJr commented on Content.Server/GameTicking/GameTicker.GameRule.cs:
```diff
@@ -11,21 +11,38 @@ public partial class GameTicker
+            RaiseLocalEvent(new GameRuleStartedEvent(rule));
+        }
+
+        public void EndGameRule(GameRulePrototype rule)
+        {
+            if (!HasGameRule(rule))
+                return;
+
+            DisableGameRule(rule);
+            RaiseLocalEvent(new GameRuleEndedEvent(rule));
+        }
```
I think at the moment someone could `Disable` a game rule that has been started, which removes it from `_gameRules` without ever ending it?

I think in the same way that starting ensures that the rule has been enabled, disabling should ensure the rule has ended, but the end-function shouldn't necessarily call disable?

Also I think you might need to distinguish between has-rule and rule-started?

mirrorcult commented:
updated this finally,

basically renamed enable/disable to add/remove again, and also removed 'removed' entirely since theres no actual reason for it to exist, any game rule that gets disabled should always be ended anyway, the only behavior that mattered was separating add/start

mirrorcult commented:
@ElectroJr

mirrorcult commented:
what the fuck is this test fail

ElectroJr commented on Content.Server/Spawners/EntitySystems/ConditionalSpawnerSystem.cs:
```diff
@@ -35,15 +31,15 @@ private void OnRandSpawnMapInit(EntityUid uid, RandomSpawnerComponent component,
         {
             foreach (var spawner in EntityManager.EntityQuery<ConditionalSpawnerComponent>())
             {
                 RuleAdded(spawner, args);
```
`RuleAdded()` should probably be renamed

ElectroJr commented on Content.Server/GameTicking/GameTicker.GameRule.cs:
```diff
@@ -47,14 +90,18 @@ public bool HasGameRule(string rule)
         public void ClearGameRules()
         {
-            foreach (var rule in _gameRules.ToArray())
+            _addedGameRules.Clear();
```
```suggestion
```
`EndGameRule(rule)` checks only ends rules that are in `_addedGameRules` and will already remove them.

Also, should this be looping over `_startedGameRules` or `_addedGameRules`?

ElectroJr commented on Content.Server/GameTicking/Rules/GameRuleSystem.cs:
```diff
@@ -37,25 +38,31 @@ private void OnGameRuleAdded(GameRuleAddedEvent ev)
+    {
+        if (ev.Rule.ID != Prototype)
+            return;
+
+        Ended();
```
missing a `Enabled = false;`, leading to infinite restart-round loops.
Probably the cause of the weird test error

mirrorcult commented:
resolved and hopefully test fail is resolved

ElectroJr reviewed (approved):
Haven't actually tested trying misc game rules with local clients, but seems all good.
